### PR TITLE
Further enhance `needless_borrow`, mildly refactor `redundant_clone`

### PR DESCRIPTION
**Pull request by smoelius**
This PR does the following:
* Moves some code from `redundant_clone` into a new `clippy_utils` module called `mir`, and wraps that code in a function called `dropped_without_further_use`.
* Relaxes the "is copyable" condition condition from #9136 by also suggesting to remove borrows from values dropped without further use. The changes involve the just mentioned function.
* Separates `redundant_clone` into modules.

Strictly speaking, the last bullet is independent of the others. `redundant_clone` is somewhat hairy, IMO. Separating it into modules makes it slightly less so, by helping to delineate what depends upon what.

I've tried to break everything up into digestible commits.

r? @Jarcho

(@Jarcho I hope you don't mind.)

changelog: continuation of #9136

### CODE REVIEW

**Comment by Jarcho** [comment]
Just did a quick check. Does the extension to `needless_borrow` handle closures? I think those have a separate MIR body from the containing function.

**Comment by smoelius** [comment]
I just add a closure-related test. Please let me know if it doesn't address your specific concern.

**Comment by Jarcho** [comment]
The use of `env` in the test could have the borrow removed (not trivially, you would have to prove the closure is only called once). It would be better to have some use of it after the closure is created.

As this suggestion changes the drop order it would be better to limit this to types without a significant drop function. You can use [`has_significant_drop`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.has_significant_drop).

**Comment by smoelius** [comment]
> It would be better to have some use of it after the closure is created.

This has revealed a bug. Please give me a day or so to look into it.

**Comment by smoelius** [comment]
Any idea what I might have done to cause this? https://github.com/rust-lang/rust-clippy/runs/8089040744?check_suite_focus=true#step:7:1476

**Comment by smoelius** [comment]
I think I figured it out. Sorry for the noise.

Today's changes move the `possible_borrower` code under `clippy_utils::mir` so that it can be used in dereference.rs.

The bug I mentioned is resolved, and the PR should be ready for review (again).

**Comment by smoelius** [comment]
Okay to rebase?

File: clippy_dev/src/new_lint.rs [hunk 1]
```diff
@@ -98,7 +98,7 @@ fn create_test(lint: &LintData<'_>) -> io::Result<()> {
         fs::create_dir(&test_dir)?;
 
         create_project_layout(lint.name, &test_dir, "fail", "Content that triggers the lint goes here")?;
-        create_project_layout(lint.name, &test_dir, "pass", "This file should not trigger the lint")?;
+        create_project_layout(lint.name, test_dir, "pass", "This file should not trigger the lint")?;
```
Comment (Jarcho) [inline clippy_dev/src/new_lint.rs:101]: This suggestion is not particularly great. It raises the question of why the first call borrows `test_dir`, but the second call doesn't. Reusing `test_dir` afterwards on a later change is now extra work as well.

**Review by Jarcho — COMMENTED** [review]
A couple of things here:
* This can calculate possible borrowers multiple times per function. It also fully recalculates the liveness bitset each time as well. What's the perf impact from this?
* Some suggestions from this are arguably downgrades (see later comment for example).

Given the second point it might be better to only lint on temporaries.

**Comment by smoelius** [comment]
> A couple of things here:
> 
> * This can calculate possible borrowers multiple times per function. It also fully recalculates the liveness bitset each time as well. What's the perf impact from this?
> * Some suggestions from this are arguably downgrades (see later comment for example).
> 
> Given the second point it might be better to only lint on temporaries.

This strategy feels a little extreme to me. For example, it would seem to eliminate a large number of valid suggestions from https://github.com/rust-lang/rust-clippy/pull/9386/commits/29a705761a03028e61eeabd47057e823c67ef7c3.

To address the first bullet, we could compute the possible borrowers once (lazily) in a `check_body` implementation, and make the current `check_expr` a method of a visitor called from `check_body` (e.g., make the `Option<PossibleBorrowMap<...>>` a field of the visitor).

To address the second bullet, may I suggest the following alternative? For each body, build a map from def ids to substitutions used in calls to that def id. (This could also be done lazily.) Then, suggest to remove a `&` from an argument in a call to `def_id` only if it would not increase the total number of substitutions used in calls to `def_id` within the body.

Whichever way we go, can we make the current, unhindered strategy something that could be optionally enabled?

**Comment by Jarcho** [comment]
Most of the changes are on temporaries (52/74) by a quick count. Could limit locals to those which are used only once, which would lint almost all cases. The main point is to avoid changing something like: 
```rust
let x = ...;
use_x(&x);
...
use_x_again(&x);
```

>To address the first bullet, we could compute the possible borrowers once (lazily) in a check_body implementation, and make the current check_expr a method of a visitor called from check_body (e.g., make the Option<PossibleBorrowMap<...>> a field of the visitor).
That would unconditionally add another traversal over every function. which would probably be worse on average. Ideally we could store this in the lint pass, but that's not possible.

**Comment by smoelius** [comment]
> That would unconditionally add another traversal over every function.

I thought what I was proposing would conditionally add another traversal (i.e., if all of the conditions were met so that the possible borrowers were needed to proceed).

Why would it be unconditional?

> Could limit locals to those which are used only once,

That sounds reasonable to me. Okay to turn this off with an option, though?

**Comment by Jarcho** [comment]
You suggested to switching the lint to use `check_body` with the current `check_expr` moved into a visitor, correct? That leaves the visitor with all the lints, plus the new visitor from `check_body` running over the code.

**Comment by smoelius** [comment]
> You suggested to switching the lint to use `check_body` with the current `check_expr` moved into a visitor, correct? That leaves the visitor with all the lints, plus the new visitor from `check_body` running over the code.

OK. I see what you are saying.

Then what if we essentially queued the arguments to `needless_borrow_impl_arg_position` inside `Position::ImplArg`, and actually processed those arguments/performed those calls in an implementation of `check_body_post`?

Admittedly, one thing I am not clear on is whether returning an `Position::ImplArg` as a possibility rather than a certainty would break the surrounding code. For example, would this code break if no warning were ultimately produced for the `ImplArg`? Maybe the answer is obvious to you.
https://github.com/rust-lang/rust-clippy/blob/5860abff70714c27daaf90b8f783812ee4da222b/clippy_lints/src/dereference.rs#L364-L365

**Comment by Jarcho** [comment]
I think you can make that work by storing the expression's `HirId` the target function's `DefId` and argument position and then running over all those in `check_body_post`. This would be slightly complicated by nested functions.

I'll check to see if we can change rustc to allow storing `TyCtxt` bounded items in lint passes as that would allow caching the MIR result directly.

**Comment by smoelius** [comment]
> I'll check to see if we can change rustc to allow storing `TyCtxt` bounded items in lint passes as that would allow caching the MIR result directly.

Thanks!

Did you already start those changes? If not, I could make them. Please just let me know.

**Comment by Jarcho** [comment]
Implemented by rust-lang/rust#101501. Hopefully that will be synced to clippy this week.

**Comment by smoelius** [comment]
Thanks, @Jarcho!

**Comment by smoelius** [comment]
I saw you PR was merged. But it looks like it didn't make it in time for the most recent rustup?

If I'm reading everything right, the current toolchain (nightly-2022-09-08) is based on commit [c2804e6](https://github.com/rust-lang/rust/commit/c2804e6ec2c29a5c7368600ea173b890e2655c3d), which is later than [8778809](https://github.com/rust-lang/rust/commit/87788097b776f8e3662f76627944230684b671bd) (the one which includes your PR).

**Comment by Jarcho** [comment]
That is correct.

**Comment by smoelius** [comment]
I think this is ready for review again, @Jarcho.

(Rebasing was a bit of bear, so hopefully I did not screw it up.)

File: clippy_lints/src/dereference.rs [hunk 2]
```diff
@@ -141,15 +146,27 @@ declare_clippy_lint! {
...
 }
 
-impl_lint_pass!(Dereferencing => [
-    EXPLICIT_DEREF_METHODS,
-    NEEDLESS_BORROW,
-    REF_BINDING_TO_REFERENCE,
-    EXPLICIT_AUTO_DEREF,
-]);
+#[expect(rustc::internal)]
+impl<'tcx> LintPass for Dereferencing<'tcx> {
```
[the remark this answers is not in the document]
Comment (smoelius) [reply clippy_lints/src/dereference.rs]: Yep. You're right.

File: clippy_lints/src/utils/internal_lints.rs [hunk 3]
```diff
@@ -426,87 +426,86 @@ impl<'tcx> LateLintPass<'tcx> for LintWithoutLintPass {
...
-            if is_deprecated_lint(cx, ty) || is_lint_ref_ty {
-                check_invalid_clippy_version_attribute(cx, item);
-
-                let expr = &cx.tcx.hir().body(body_id).value;
-                let fields;
-                if is_lint_ref_ty {
-                    if let ExprKind::AddrOf(_, _, inner_exp) = expr.kind
-                        && let ExprKind::Struct(_, struct_fields, _) = inner_exp.kind {
-                            fields = struct_fields;
+        match &item.kind {
```
[the remark this answers is not in the document]
Comment (smoelius) [reply clippy_lints/src/utils/internal_lints.rs]: These changes seemed to be necessary to suppress `lint_without_lint_pass` when I was implementing `LintPass` manually. But as you noted, implementing `LintPass` manually is not necessary.

I have reverted these changes.

**Comment by smoelius** [comment]
I think the three most recent comments have been addressed.

**Comment by Jarcho** [comment]
I think the only thing remaining is the previous point about changing `f1(&x); f2(&x);` to `f1(&x); f2(x);`

**Comment by smoelius** [comment]
> I think the only thing remaining is the previous point about changing `f1(&x); f2(&x);` to `f1(&x); f2(x);`

Meaning _don't_ suggest that, right?

It's possible there's a problem, but I think I addressed that, actually, e.g.:
* https://github.com/rust-lang/rust-clippy/blob/252f598f5a850bde9026cf3582a54e1c2423a230/clippy_lints/src/dereference.rs#L1126
* https://github.com/rust-lang/rust-clippy/blob/252f598f5a850bde9026cf3582a54e1c2423a230/tests/ui/needless_borrow.rs#L367-L383
* https://github.com/rust-lang/rust-clippy/blob/252f598f5a850bde9026cf3582a54e1c2423a230/tests/ui/needless_borrow.stderr#L207-L211

**Comment by Jarcho** [comment]
So it is. I was going by the change in `clippy_dev/src/new_lint.rs` still being there. (there's no need to change this).

You can rebase and squash.

**Comment by smoelius** [comment]
> So it is. I was going by the change in `clippy_dev/src/new_lint.rs` still being there. (there's no need to change this).

It should have occurred to me to fix that.

I will fix that and any other similar changes before rebasing.

**Comment by smoelius** [comment]
The most recent push squashes all of the commits except for the "Fixed adjacent code" ones.

The current "Fixed adjacent code" commit is the result of rerunning the lint.

Sadly, the introduction of `possible_borrowers` eliminated more suggestions than I realized.

I think there may be ways `possible_borrowers` could be improved, but that could be the subject of a separate PR.

**Comment by Jarcho** [comment]
Everything looks good. If you can catch more single use cases that would be great.

Thank you. @bors r+

**Comment by smoelius** [comment]
> Thank you. @bors r+

Thank you as well! Especially for https://github.com/rust-lang/rust/pull/101501. :pray:

Cheers.

**Pull request merged**